goldmane/pkg/aggregator/aggregator.go

```diff
@@ -0,0 +1,307 @@
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggregator
+
+import (
+	"sort"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/goldmane/pkg/internal/types"
+	"github.com/projectcalico/calico/goldmane/proto"
+)
+
+const (
+	// numBuckets is the number of buckets to keep in memory.
+	// We keep 120 buckets. Assuming a default window of 15s each, this
+	// gives us a total of 30 minutes of history.
+	numBuckets = 120
+
+	// channelDepth is the depth of the channel to use for flow updates.
+	channelDepth = 5000
+)
+
+// Sink is an interface that can receive aggregated flows.
+type Sink interface {
+	Receive(*AggregationBucket)
+}
+
+// flowRequest is an internal helper used to synchronously request matching flows from the aggregator.
+type flowRequest struct {
+	respCh chan []proto.Flow
+	req    *proto.FlowRequest
+}
+
+type LogAggregator struct {
+	buckets []AggregationBucket
+
+	// The rollover time defines how often we rollover to a new bucket.
+	rolloverTime time.Duration
+
+	// Used to trigger goroutine shutdown.
+	done chan struct{}
+
+	// Used to make requests for flows synchronously.
+	flowRequest chan flowRequest
+
+	// sink is a sink to send aggregated flows to.
+	sink Sink
+
+	// recvChan is the channel to receive flow updates on.
+	recvChan chan *proto.FlowUpdate
+
+	// rolloverFunc allows manual control over the rollover timer, used in tests.
+	// In production, this will be time.After.
+	rolloverFunc func(time.Duration) <-chan time.Time
+
+	// bucketsToAggregate is the number of internal buckets to aggregate when pushing flows to the sink.
+	// This can be used to reduce the number of distinct flows that are sent to the sink, at the expense of
+	// delaying the emission of flows.
+	// 20 buckets of 15s provides a 5 minute aggregation.
+	bucketsToAggregate int
+
+	// pushIndex is the index of the bucket to push to the emitter. We only push
+	// the bucket after it has been rolled over several times, to ensure that we have
+	// a complete view of the flows in the bucket.
+	//
+	// Increasing this value will increase the latency of the emitted flows. Decreasing it too much
+	// will cause the emitter to emit incomplete flows.
+	//
+	// Latency-to-emit is roughly (pushIndex * rolloverTime).
+	pushIndex int
+}
+
+func NewLogAggregator(opts ...Option) *LogAggregator {
+	// Establish default aggregator configuration.
+	a := &LogAggregator{
+		rolloverTime:       1 * time.Minute,
```

Comment:
If we want 5min aggregation with 20 buckets, shouldn't this be 15sec?

Reply:
Yeah, we actually override this default value with one in cmd/main.go... but it's a bit confusing to have a non-functional default here. I'll update to match what we actually use.